### PR TITLE
Estimate inner join row count using a middle ground method

### PR DESCRIPTION
**Pull request by Youngwb**
#535

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java [hunk 1]
```diff
@@ -871,10 +875,20 @@ public double estimateInnerRowCount(Statistics statistics, List<BinaryPredicateO
...
-        // "driving predicate" which mostly reduces join output rows cardinality and apply UNKNOWN_FILTER_COEFFICIENT
-        // for other (auxiliary) predicates.
+        if (ConnectContext.get().getSessionVariable().isUseCorrelatedJoinEstimate()) {
+            return estimatedInnerRowCountAssumeCorrelated(statistics, eqOnPredicates);
+        } else {
+            return estimateInnerRowCountMiddleGround(statistics, eqOnPredicates);
+        }
+    }
+
+    // The implementation here references Presto
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:885]:
```suggestion
    // The implementation here refers to Presto
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java [hunk 2]
```diff
@@ -890,6 +904,81 @@ public double estimateInnerRowCount(Statistics statistics, List<BinaryPredicateO
...
+    private double getPredicateSelectivity(PredicateOperator predicateOperator, Statistics statistics) {
+        Statistics estimatedStatistics = estimateStatistics(Lists.newArrayList(predicateOperator), statistics);
+        return estimatedStatistics.getOutputRowCount() / statistics.getOutputRowCount();
+    }
+
+    //  This estimate join row count method references ORCA.
+    //  use a damping method to moderately decrease the impact of subsequent predicates to account for correlated columns.
+    //  This damping only occurs on sorted predicates of the same table, otherwise we assume independence.
+    //  For example, given ANDed predicates (t1.a = t2.a AND t1.b = t2.b AND t2.b = t3.a) with the given selectivity:
+    //  (S1) t1.a = t2.a has selectivity 0.3
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:916]: What's the meaning of `S`?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java [hunk 3]
```diff
@@ -890,6 +904,81 @@ public double estimateInnerRowCount(Statistics statistics, List<BinaryPredicateO
...
+
+            ColumnRefSet leftChildColumns = predicateOperator.getChild(0).getUsedColumns();
+            ColumnRefSet rightChildColumns = predicateOperator.getChild(1).getUsedColumns();
+            List<Integer> leftChildRelationIds =
+                    leftChildColumns.getStream().mapToObj(columnRefFactory::getRelationId).collect(Collectors.toList());
+            List<Integer> rightChildRelationIds =
+                    rightChildColumns.getStream().mapToObj(columnRefFactory::getRelationId)
+                            .collect(Collectors.toList());
+
+            if (leftChildRelationIds.size() == 1 && rightChildRelationIds.size() == 1) {
```
[the remark this answers is not in the document]
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:969]: It's used to check that the predicate is complex,  such as t1.a + t2.b = t3.c is complex predicate, we need relation id

[the remark this answers is not in the document]
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:969]: Fixed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java (hunk 2 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:917]: `t1.a = t2.a` is a predicate, not a selectivity?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 4]
```diff
@@ -161,6 +161,7 @@
     public static final String CBO_MAX_REORDER_NODE_USE_DP = "cbo_max_reorder_node_use_dp";
     public static final String CBO_ENABLE_GREEDY_JOIN_REORDER = "cbo_enable_greedy_join_reorder";
     public static final String CBO_ENABLE_REPLICATED_JOIN = "cbo_enable_replicated_join";
+    public static final String USE_CORRELATED_JOIN_ESTIMATE = "use_correlated_join_estimate";
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:154]: OK, CBO session variables should start with "cbo_"

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java (hunk 2 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java]: moved the S position

**Review by kangkaisen — DISMISSED** [review]

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**